### PR TITLE
add adjoint definitions

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
Derivatives at the boundary are only setup for 1-D right now: they are incorrect for higher dimension but that can be handled in a next PR.

### CODE REVIEW

File: src/Quadrature.jl [hunk 1]
```diff
@@ -372,6 +390,66 @@ function __init__()
...
+                    back(y)[1]
+                end
+            elseif sensealg.vjp isa ReverseDiffVJP
+                error("TODO")
+            end
+        end
+        dlb = -_f(lb)
+        dub = _f(ub)
+
+        dp_prob = QuadratureProblem(dfdp,lb,ub,p;nout=length(p),batch = prob.batch,kwargs...)
```
Comment (YingboMa) [inline src/Quadrature.jl:435]: It seems that the boundary term is missing. Will Zygote handle parameter-dependent `lb` and `ub` automatically?

Comment (ChrisRackauckas) [reply src/Quadrature.jl:435]: yes, as long as this definition for the derivative at the boundary is correct. So for 1-D this already works. For N-D I need to think more (Stokes Theorem, etc.)

Comment (ChrisRackauckas) [reply src/Quadrature.jl:436]: I set it to return `nothing` in the cases it's incorrect for, for now. Our applications don't need this derivative anyways.

File: src/Quadrature.jl [hunk 2]
```diff
@@ -106,15 +127,15 @@ function DiffEqBase.solve(prob::QuadratureProblem,alg::VEGAS,args...;
...
     DiffEqBase.build_solution(prob,alg,val,err,chi=chi,retcode = :Success)
 end
 
 function __init__()
     @require Cubature="667455a9-e2ce-5579-9412-b964f529a492" begin
-        function DiffEqBase.solve(prob::QuadratureProblem,
+        function __solvebp_call(prob::QuadratureProblem,
                                   alg::AbstractCubatureJLAlgorithm, args...;
                                   reltol = 1e-8, abstol = 1e-8,
                                   maxiters = typemax(Int),
```
Comment (agerlach) [inline src/Quadrature.jl:141]: Missing `sensealg,lb, ub, p` 

```julia
function __solvebp_call(prob::QuadratureProblem,
                                  alg::AbstractCubatureJLAlgorithm, sensealg,
                                  lb, ub, p, args...;
                                  reltol = 1e-8, abstol = 1e-8,
                                  maxiters = typemax(Int),
                                  kwargs...)
```

File: test/runtests.jl [hunk 3]
```diff
@@ -1,311 +1,6 @@
-##
-using Quadrature
-using Cubature, Cuba
+using Pkg
```
Comment (agerlach) [inline test/runtests.jl:1]: Stray usage. `Pkg` not in Project.toml

File: test/runtests.jl [hunk 4]
```diff
@@ -1,311 +1,6 @@
...
+using SafeTestsets
```
Comment (agerlach) [inline test/runtests.jl:2]: Needs added to Project.toml

File: test/derivative_tests.jl [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+sol = solve(prob,CubaCuhre(),reltol=1e-3,abstol=1e-3)
+
+function testf(lb,ub,p)
+    prob = QuadratureProblem(f,lb,ub,p,nout=2)
+    sum(solve(prob,CubaCuhre(),reltol=1e-3,abstol=1e-3))
+end
+dlb1,dub1,dp1 = Zygote.gradient(testf,lb,ub,p)
+dlb2 = FiniteDiff.finite_difference_gradient(lb->testf(lb,ub,p),lb)
+dub2 = FiniteDiff.finite_difference_gradient(ub->testf(lb,ub,p),ub)
+dp2 = FiniteDiff.finite_difference_gradient(p->testf(lb,ub,p),p)
```
Comment (agerlach) [inline test/derivative_tests.jl:76]: The `dp` solutions don't look correct. Returns a 2 element vector. w/ `nout=2` shouldn't the solution be 2x2 or a vec of vec?

Comment (ChrisRackauckas) [reply test/derivative_tests.jl:76]: I'm taking the sum of the result

Comment (agerlach) [reply test/derivative_tests.jl:98]: I see... that makes sense.

**Comment by ChrisRackauckas** [comment]
The PR passes locally now, but for some reason something is going on on Travis. I'll need to do some version checking if this doesn't pass again, but it should be usable now.

**Comment by agerlach** [comment]
I'm seeing some odd behaviour. Tests pass locally but
```
f(x,p) = sin(x)
qprob = QuadratureProblem(f,0.0,1.0)
solve(qprob,CubatureJLh())
```
gives
```
MethodError: no method matching __solvebp_call(::QuadratureProblem{true,DiffEqBase.NullParameters,DiffEqUncertainty.var"#10#18"{Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},var"#31#32",Array{Uniform{Float64},1},Array{Uniform{Float64},1},ODEProblem{Array{Float64,1},Tuple{Float64,Float64},true,Array{Float64,1},ODEFunction{true,typeof(eom),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},DiffEqBase.StandardODEProblem},Tuple{Tsit5,EnsembleThreads},Int64},Array{Float64,1},Array{Float64,1},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}}, ::CubatureJLh, ::Quadrature.ReCallVJP{Quadrature.ZygoteVJP}, ::Array{Float64,1}, ::Array{Float64,1}, ::DiffEqBase.NullParameters; reltol=0.01, abstol=0.01, maxiters=0)
Closest candidates are:
  __solvebp_call(::QuadratureProblem, !Matched::QuadGKJL, ::Any, ::Any, ::Any, ::Any, !Matched::Any...; reltol, abstol, maxiters, kwargs...) at /Users/gerlacar/.julia/packages/Quadrature/rET3h/src/Quadrature.jl:70
  __solvebp_call(::QuadratureProblem, !Matched::HCubatureJL, ::Any, ::Any, ::Any, ::Any, !Matched::Any...; reltol, abstol, maxiters, kwargs...) at /Users/gerlacar/.julia/packages/Quadrature/rET3h/src/Quadrature.jl:87
  __solvebp_call(::QuadratureProblem, !Matched::VEGAS, ::Any, ::Any, ::Any, ::Any, !Matched::Any...; reltol, abstol, maxiters, kwargs...) at /Users/gerlacar/.julia/packages/Quadrature/rET3h/src/Quadrature.jl:113
__solvebp(::QuadratureProblem{true,DiffEqBase.NullParameters,DiffEqUncertainty.var"#10#18"{Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},var"#31#32",Array{Uniform{Float64},1},Array{Uniform{Float64},1},ODEProblem{Array{Float64,1},Tuple{Float64,Float64},true,Array{Float64,1},ODEFunction{true,typeof(eom),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},DiffEqBase.StandardODEProblem},Tuple{Tsit5,EnsembleThreads},Int64},Array{Float64,1},Array{Float64,1},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}}, ::Vararg{Any,N} where N; kwargs::Base.Iterators.Pairs{Symbol,Real,Tuple{Symbol,Symbol,Symbol},NamedTuple{(:reltol, :abstol, :maxiters),Tuple{Float64,Float64,Int64}}}) at Quadrature.jl:64
(::Quadrature.var"#__solvebp##kw")(::NamedTuple{(:reltol, :abstol, :maxiters),Tuple{Float64,Float64,Int64}}, ::typeof(Quadrature.__solvebp), ::QuadratureProblem{true,DiffEqBase.NullParameters,DiffEqUncertainty.var"#10#18"{Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},var"#31#32",Array{Uniform{Float64},1},Array{Uniform{Float64},1},ODEProblem{Array{Float64,1},Tuple{Float64,Float64},true,Array{Float64,1},ODEFunction{true,typeof(eom),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},DiffEqBase.StandardODEProblem},Tuple{Tsit5,EnsembleThreads},Int64},Array{Float64,1},Array{Float64,1},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}}, ::CubatureJLh, ::Quadrature.ReCallVJP{Quadrature.ZygoteVJP}, ::Array{Float64,1}, ::Array{Float64,1}, ::DiffEqBase.NullParameters) at Quadrature.jl:64
solve(::QuadratureProblem{true,DiffEqBase.NullParameters,DiffEqUncertainty.var"#10#18"{Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},var"#31#32",Array{Uniform{Float64},1},Array{Uniform{Float64},1},ODEProblem{Array{Float64,1},Tuple{Float64,Float64},true,Array{Float64,1},ODEFunction{true,typeof(eom),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},DiffEqBase.StandardODEProblem},Tuple{Tsit5,EnsembleThreads},Int64},Array{Float64,1},Array{Float64,1},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}}, ::CubatureJLh; sensealg::Quadrature.ReCallVJP{Quadrature.ZygoteVJP}, kwargs::Base.Iterators.Pairs{Symbol,Real,Tuple{Symbol,Symbol,Symbol},NamedTuple{(:reltol, :abstol, :maxiters),Tuple{Float64,Float64,Int64}}}) at Quadrature.jl:60
(::DiffEqBase.var"#solve##kw")(::NamedTuple{(:reltol, :abstol, :maxiters),Tuple{Float64,Float64,Int64}}, ::typeof(solve), ::QuadratureProblem{true,DiffEqBase.NullParameters,DiffEqUncertainty.var"#10#18"{Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},var"#31#32",Array{Uniform{Float64},1},Array{Uniform{Float64},1},ODEProblem{Array{Float64,1},Tuple{Float64,Float64},true,Array{Float64,1},ODEFunction{true,typeof(eom),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}...
```

**Comment by ChrisRackauckas** [comment]
Cubature is in requires:

```julia
using Quadrature, Cubature
f(x,p) = sin(x)
qprob = QuadratureProblem(f,0.0,1.0)
solve(qprob,CubatureJLh())
```

**Comment by agerlach** [comment]
> Cubature is in requires:

Thats embarrassing... carry on

**Pull request merged**